fly/src/sorts.rs

```diff
@@ -528,92 +728,126 @@ impl Context<'_> {
+    }
+
+    // doesn't allow `binders` to shadow each other, but does allow them to
+    // shadow names already in `context`
+    //
+    // for any variables that do not have a sort annotation, this function allocates
+    // a fresh unification variable to represent its sort, and annotates the AST with
+    // a string that uniquely identifies the unification variable. unification variables
+    // are represented by integers, and the string "var 55" is used to represent, eg, the
+    // unification variable numbered 55. since this string has a space in it, it is impossible
+    // for it to be confused with a user sort annotation.
+    fn add_binders_for_inference(&mut self, binders: &mut [Binder]) -> Result<(), SortError> {
+        let mut names = HashSet::new();
+        for binder in binders {
+            // First check that the name is not repeated *within* this slice.
+            if !names.insert(binder.name.clone()) {
+                return Err(SortError::RedeclaredName(binder.name.clone()));
             }
-            Term::NAryOp(NOp::And | NOp::Or, xs) => {
-                for x in xs {
-                    let sort = self.sort_of_term(x)?;
-                    self.unify_var_value(&Sort::Bool, &sort)?;
-                }
-                Ok(AbstractSort::Known(Sort::Bool))
+            self.scope.check_sort_exists_or_empty(&binder.sort)?;
+            let sort = if binder.sort == Sort::Id("".to_owned()) {
+                let sort_var = self.new_key();
+                binder.sort = Sort::Id(format!("var {}", sort_var.0));
+                RelationOrIndividual::unknown(sort_var)
+            } else {
+                RelationOrIndividual::known(&binder.sort)
+            };
+            // Now add it to the context, allowing it to shadow bindings from
+            // any outer scopes.
+            self.scope
+                .add_name_internal(binder.name.clone(), sort, ShadowingConstraint::Allow)?;
+        }
+        Ok(())
+    }
+
+    /// Check the given value sort with the given variable known-or-unknown sort.
+    /// If the variable sort is known, check if its solution is equal, and if not, report an error.
+    /// If the variable sort is unknown, resolve the variable to the given value sort.
+    fn unify_var_value(&mut self, value: &Sort, var: &MaybeUnknownSort) -> Result<(), SortError> {
+        self.unification_table.unify_var_value(value, var)
+    }
+
+    /// Check the two sort variables have the same known solution.
+    /// Or, if at least one is unknown, unify the two.
+    fn unify_var_var(
+        &mut self,
+        a: &MaybeUnknownSort,
+        b: &MaybeUnknownSort,
+    ) -> Result<(), SortError> {
+        self.unification_table.unify_var_var(a, b)
+    }
+}
+
+/// Thin wrapper around ena::unify::UnificationTable that supports a few convenience methods for unification
+struct UnificationTable {
+    inner: ena::unify::InPlaceUnificationTable<SortVar>,
+}
```

Comment:
I don't think we need this wrapper; I would prefer to put the code and docs for methods for `UnificationTable` into the methods of the same name for `InternalContext`. If we do keep this wrapper, we should remove the methods in `InternalContext`.